### DISCUSSION TITLE
Generics in Setup

### DISCUSSION THREAD

**Discussion by AngeloHulshout** [reqnroll/Reqnroll#376] [Q&A] [answered]
I am migrating from SpecFlow to Reqnroll - understandably in a relative hurry. I chose to bypass the SpecFlow compatibiliy assembly and just rename Techtalk.Specflow to Reqnroll. That surprisingly made everything build without problems right away. 
However, I have some generics in the setup stage, and when running my tests I am confronted with "Reqnroll.BindingException : Binding error(s) found:  Binding types cannot be generic: "

That is something we can refactor around, but it brings two questions:

1. Would this work with the Specflow compatibility assembly? No need to answer, I'll give it a try.
2. Why are generics not supported in reqnroll during setup? Is this a temporary limitation or a decisison towards the future?

And apart from that: thanks for the fork, this is going to save us a lot of trouble now that SpecFlow will be gone in 3 days.

**Comment by olegKoshmeliuk** [comment]
Hi,

Looks like you trying to use open generics in bindings.
Since open generic types can't created without type parameters that's why you getting an error.
Maybe you by mistake put [Binding] on abstract generic class?

Ideally add example proejct to get an idea where you using open generic types in Specflow or Reqnroll.
```cs
[Binding]  // this will throw you error
public class StepsBase<T> { }

[Binding]
public class UserSteps: StepsBase<User> { }

```


code related to your error: https://github.com/reqnroll/Reqnroll/blob/642c8bcfc4959de18308705ed8a7800ca353e877/Reqnroll/Bindings/Discovery/BindingSourceProcessor.cs#L311

**Reply by AngeloHulshout** [reply]
Thank you for that link. This is definitely related, but I need to do some refactoring as I said. Still don't understand why this does not work with reqnroll while it did with SpecFlow, but we'll sort that out.  In line with your example, what I have is:

[Binding]
public class TestBase
...

abstract class TestBase<TSetupService> : TestBase where TSetupService : class, ISetupService
....

Binding types cannot be generic: app.specs.TestBase`1[TSetupService]

**Comment by olegKoshmeliuk** [comment]
Can you show example project with valid code? (just create specflow/reqnroll empty project) and edit CalculatorStepDefinitions.
Even in specflow  startup project don't allow generic types for Bindings. Tried step bindings and BeforeScenario hooks

![image](https://github.com/user-attachments/assets/08476d51-69ff-4e20-873c-03d4403ce7f7)
![image](https://github.com/user-attachments/assets/09dd2fb5-761b-4e4a-a8d0-ba25f7bb37ef)


BTW original specflow have this check too. 
https://github.com/endjin/SpecFlow/blob/d93324577d603a7096d264b4fb60a8e49a07919c/TechTalk.SpecFlow/Bindings/Discovery/BindingSourceProcessor.cs#L243

**Comment by AngeloHulshout** [comment]
I believe you, Oleg, but ... my original project compiled and ran fine until klast week with SpecFlow 3.9 - after switching to reqnroll this problem started occurring. 

My previous post didn't show correclty, for some reason the discussion software removed the 'generic' indication.

This is what I had:

```
[Binding]
public class TestBase {
...
}

abstract class TestBase<TSetupService> : TestBase where TSetupService : class, ISetupService
{
...
}


class PTest : TestBase<IPSetup>
{
...
}

class SingleTest : PTest
{
...
}
```


I fixed the problem last night by doing this - moving the Binding to the concrete test. No idea how this always worked, we've been using it like above for over a year. 

```
public class TestBase {
...
}

abstract class TestBase<TSetupService> : TestBase where TSetupService : class, ISetupService
{
...
}


class PTest : TestBase<IPSetup>
{
...
}

[Binding]
class SingleTest : PTest
{
...
}
```

**Reply by olegKoshmeliuk** [reply, marked as answer]
It's new behaviour of reqnroll (maybe it was planned for specflow 4.0). Reqnroll validates all bindings inside `registry` and in case of invalid fails early. 
Binding attribute is applied to base and all child classes.
For you case you need to move `Binding` attribute from base to child classes that will be created by Reqnroll during steps.

**Reply by AngeloHulshout** [reply]
That's basically what I did. Thanks for guiding the thought process!
